homeassistant/components/mqtt/device_trigger.py

```diff
@@ -0,0 +1,277 @@
+    {
+        vol.Required(CONF_PLATFORM): DEVICE,
+        vol.Required(CONF_DOMAIN): DOMAIN,
+        vol.Required(CONF_DEVICE_ID): str,
+        vol.Required(CONF_DISCOVERY_ID): str,
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    }
+)
+
+TRIGGER_DISCOVERY_SCHEMA = mqtt.MQTT_BASE_PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_AUTOMATION_TYPE): str,
+        vol.Required(CONF_DEVICE): mqtt.MQTT_ENTITY_DEVICE_INFO_SCHEMA,
+        vol.Required(CONF_TOPIC): mqtt.valid_subscribe_topic,
+        vol.Optional(CONF_PAYLOAD, default=None): vol.Any(None, cv.string),
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_SUBTYPE): cv.string,
+    },
+    mqtt.validate_device_has_at_least_one_identifier,
+)
+
+DEVICE_TRIGGERS = "mqtt_device_triggers"
+ATTACHED_DEVICE_TRIGGERS = "mqtt_attached_device_triggers"
+
+
+@attr.s(slots=True)
+class Trigger:
+    """Device trigger settings."""
+
+    device_id = attr.ib(type=str)
+    discovery_id = attr.ib(type=str)
+    hass = attr.ib(type=HomeAssistantType)
+    payload = attr.ib(type=str)
+    qos = attr.ib(type=int)
+    subtype = attr.ib(type=str)
+    topic = attr.ib(type=str)
+    type = attr.ib(type=str)
+
+    async def attach_trigger(self, action, automation_info):
+        """Attach MQTT trigger."""
+        mqtt_config = {
+            automation_mqtt.CONF_TOPIC: self.topic,
+            automation_mqtt.CONF_ENCODING: DEFAULT_ENCODING,
+            automation_mqtt.CONF_QOS: self.qos,
+        }
+        if self.payload:
+            mqtt_config[CONF_PAYLOAD] = self.payload
+
+        return await automation_mqtt.async_attach_trigger(
+            self.hass, mqtt_config, action, automation_info
+        )
+
+    async def update_trigger(self, config):
+        """Update MQTT device trigger."""
+        self.type = config[CONF_TYPE]
+        self.subtype = config[CONF_SUBTYPE]
+        self.topic = config[CONF_TOPIC]
+        self.payload = config[CONF_PAYLOAD]
+        self.qos = config[CONF_QOS]
+
+        if ATTACHED_DEVICE_TRIGGERS not in self.hass.data:
+            return
+
+        # Unsubscribe+subscribe if this trigger is in use
+        for trig in self.hass.data[ATTACHED_DEVICE_TRIGGERS]:
+            if trig.discovery_id == self.discovery_id:
+                if trig.remove:
+                    trig.remove()
+                trig.remove = await self.attach_trigger(
+                    trig.action, trig.automation_info
+                )
+
+    def detach_trigger(self):
+        """Remove MQTT device trigger."""
+
+        # Unsubscribe if this trigger is in use
+        for trig in self.hass.data[ATTACHED_DEVICE_TRIGGERS]:
+            if trig.discovery_id == self.discovery_id:
+                if trig.remove:
+                    trig.remove()
+
+
+@attr.s(slots=True)
+class AttachedTrigger:
+    """Attached trigger settings."""
+
+    device_id = attr.ib(type=str)
```

Comment:
Why would this need to be tracked ? Shouldn't we just keep track of remove, remove and automation_config ?

Reply:
Removed.